R/powerEmp.R

```diff
@@ -0,0 +1,155 @@
+  assert_list(simTrials, null.ok = FALSE)
+  assert_positive_number(critical)
+  assert_count(eventNum, positive = TRUE)
+
+  # Censor simulated trials at time-point of OS/PFS analysis.
+  trialsAna <- lapply(
+    simTrials, censoringByNumberEvents,
+    eventNum, typeEvent
+  )
+  # Compute log-rank test for all trials for OS/PFS.
+  passedTests <- unlist(lapply(
+    trialsAna,
+    logRankTest,
+    typeEvent,
+    critical
+  ))
+  passedTests
+}
+
+#' Empirical Power for a List of Simulated Trials
+#'
+#' This function computes four types of empirical power — PFS, OS, at-least (significant
+#' in at least one of PFS/OS), and joint (significant in both PFS and OS) — using
+#' the log-rank test. Empirical power is calculated as the proportion of significant
+#' results in simulated trials, each ending when a set number of PFS/OS events occur.
+#' Critical values for PFS and OS test significance must be specified.
+#'
+#' @param simTrials (`list`)\cr simulated trial data sets, see [getClinicalTrials()].
+#' @param criticalPFS (positive `number`)\cr critical value of the log-rank test for PFS.
+#' @param criticalOS (positive `number`)\cr critical value of the log-rank test for OS.
+#' @param eventNumPFS (`integer`)\cr number of PFS events required to trigger PFS analysis.
+#' @param eventNumOS (`integer`)\cr number of OS events required to trigger OS analysis.
+#'
+#' @return This returns values of four measures of empirical power.
+#' @export
+#'
+#' @examples
+#' library(survival)
+#' transition1 <- exponential_transition(h01 = 0.06, h02 = 0.3, h12 = 0.3)
+#' transition2 <- exponential_transition(h01 = 0.1, h02 = 0.4, h12 = 0.3)
+#' simTrials <- getClinicalTrials(
+#'   nRep = 50, nPat = c(800, 800), seed = 1234, datType = "1rowPatient",
+#'   transitionByArm = list(transition1, transition2), dropout = list(rate = 0.5, time = 12),
+#'   accrual = list(param = "intensity", value = 7)
+#' )
+#' powerEmp(
+#'   simTrials = simTrials, criticalPFS = 2.4, criticalOS = 2.2,
+#'   eventNumPFS = 300, eventNumOS = 500
+#' )
+powerEmp <- function(simTrials, criticalPFS, criticalOS, eventNumPFS, eventNumOS) {
+  assert_list(simTrials, null.ok = FALSE)
+  assert_positive_number(criticalPFS)
+  assert_positive_number(criticalOS)
+  assert_count(eventNumPFS, positive = TRUE)
+  assert_count(eventNumOS, positive = TRUE)
+
+  nRep <- length(simTrials)
+  simTrials <- lapply(simTrials, function(x) if (ncol(x) == 9) getDatasetWideFormat(x) else x)
+
+  # Which trials passed the log-rank test for PFS/OS?
+  passedLogRankPFS <- passedLogRank(
+    simTrials = simTrials,
+    typeEvent = "PFS",
+    eventNum = eventNumPFS,
+    critical = criticalPFS
+  )
+  passedLogRankOS <- passedLogRank(
+    simTrials = simTrials,
+    typeEvent = "OS",
+    eventNum = eventNumOS,
+    critical = criticalOS
+  )
+
+  # Empirical power is the fraction of trials with significant log-rank test:
+  powerPFS <- sum(passedLogRankPFS) / nRep
```

Comment:
So this is also the same for both, would it make sense to pull this part into the helper function too? And just let it return the power?

Reply:
I guess in this case it makes sense this way since we need the explicit list of which trials passed/failed to compute powerAtLeast/powerJoint in the next step.

Reply:
Ah right, sorry did not see that yesterday- agree to keep as is then